File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* flattenRecuresively(TreeNode* root){
        if(!root || (!root->right && !root->left))
            return root;

        TreeNode* rightChild = nullptr;
        if(root->left){
            rightChild = root->right;
            root->right = root->left;
            root->left = nullptr;
        }

        TreeNode* next = flattenRecuresively(root->right);
        if(rightChild){
            next->right = rightChild;
            return flattenRecuresively(next->right);
        }
        else
            return next;
    }
    void flatten(TreeNode* root) {
       flattenRecuresively(root);
    }
};

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def handle_subtree(self, root: Optional[TreeNode]) -> TreeNode:
        right = root.right
        if root.left:
            root.right = root.left
            root.left = None
            root = self.handle_subtree(root.right)
        if right:
            root.right = right
            root = self.handle_subtree(right)

        return root

    def flatten(self, root: Optional[TreeNode]) -> None:
        if root:
            self.handle_subtree(root)
        